### PR TITLE
Add expect call hint local

### PR DESCRIPTION
**Pull request by karol-bisztyga**

### CODE REVIEW

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (Arcticae) [inline protostar/cheatable_starknet/controllers/expect_call_controller.py:47]: This one is weird, since i think there are no side effects to instantiation of controllers anywhere else. I think this should be exposed as a method and called elsewhere. It's going to work as-is but from my perspective the disctintion between the controller and test control flow is now getting a bit blurry.

File: protostar/cheatable_starknet/callable_hint_locals/call_hint_local.py
Comment (Arcticae) [inline protostar/cheatable_starknet/callable_hint_locals/call_hint_local.py:60]:
```suggestion
                    calldata=calldata or [],
```

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (Arcticae) [inline protostar/cheatable_starknet/controllers/expect_call_controller.py:29]: it could be implemented as a `__eq__` method for `ExpectedCall` dataclass

Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py:29]: If we could have like a `FunctionCall` (name up to you) class which contains `selector` and `calldata`, we could store it in a map as a value.

Then the structure would look like this:
```
@dataclass
class FunctionCall:
    fn_selector: Selector
    calldata: CairoOrPythonData
    def __eq__(self, other):
        ... # Compare selector & calldata

@dataclass
class ExpectedCall:
    address: Address
    call: FunctionCall
    def __eq__(self, other):
        ... # Compare call and address
```

And you can then iterate over expected calls for address and compare with a `==` operator.

Also: please consider renaming `data_for_address` since it does not indicate what the data actually is. My proposition: `expected_calls` (leaving final decision up to you).

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (Arcticae) [inline protostar/cheatable_starknet/controllers/expect_call_controller.py:59]: I think i've already mentioned this before and i might have forgotten why we don't use default dicts as an implementation for those maps?

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (Arcticae) [inline protostar/cheatable_starknet/controllers/expect_call_controller.py:61]:
```suggestion
    def stop_expecting_call(self, expected_call: ExpectedCall):
```

File: protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py
Comment (Arcticae) [inline protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py:22]: The name of this hintlocal is misleading, since it _asserts_ - not removes the expectation

Comment (Arcticae) [reply protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py:22]: Also, this will be kind of cumbersome to use, since we have to pass all the data again. It'd be better if we returned an `int` (as an assertion `id`) and this hintlocal would act upon it. We would probably need to store things differently, or store additional data because we need to retrieve the expectation later by the returned `id`, but i don't think it'd be an issue to have some wrappers for manipulating contents of this map (please strongly consider using only one map if possible, since synchronizing 2 maps state would be a headache).

Comment (Arcticae) [reply protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py:22]: moreover, above can be implemented in a separate PR, not to stop this one from merging

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (Arcticae) [inline protostar/cheatable_starknet/controllers/expect_call_controller.py:22]: what does the name mean?

File: tests/integration/pure_cairo_vm/cheatcodes/expect_call/expect_call_cairo_cheatcode_test.py
Comment (Arcticae) [inline tests/integration/pure_cairo_vm/cheatcodes/expect_call/expect_call_cairo_cheatcode_test.py:1]: Very good comprehensive tests. Respect 🔥

**Review by Arcticae — CHANGES_REQUESTED** [review]
Can you please divide the 2 introduced hint locals into 2 separate PRs? 
Apart from that and the nitpicking, this looks really good :)

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py:29]: good idea 👍

File: protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py:22]: I think we should add this separately

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py:47]: Discussed offline, as it is the same as in https://github.com/software-mansion/protostar/blob/735ac6cac0a9053348ab33dc8758c2520cb2dc09/protostar/cheatable_starknet/controllers/expect_events_controller.py#L69, we decided to do this separately: https://github.com/software-mansion/protostar/issues/1500

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py:59]: https://github.com/software-mansion/protostar/pull/1363#discussion_r1068289098

File: protostar/cheatable_starknet/callable_hint_locals/call_hint_local.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/callable_hint_locals/call_hint_local.py]: [change remove expected call from external to static](https://github.com/software-mansion/protostar/pull/1481/commits/1179d77e4e82e4d969ac27da7712c09d39001975)

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py]: [change remove expected call from external to static](https://github.com/software-mansion/protostar/pull/1481/commits/1179d77e4e82e4d969ac27da7712c09d39001975)

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py]: [improve types for expected calls](https://github.com/software-mansion/protostar/pull/1481/commits/05258cda82feed9101e5989c605328b565091a41)

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py]: Deferred to #1501

**Comment by karol-bisztyga** [comment]
> Can you please divide the 2 introduced hint locals into 2 separate PRs? Apart from that and the nitpicking, this looks really good :)

Ok, I split it into two parts, here we have expect call and the part for stop expect call is in https://github.com/software-mansion/protostar/pull/1501

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py:16]: It allows us to store it as a value in a map without the map being redundant. (I recommended this solution)

Comment (MaksymilianDemitraszek) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py:16]: Yes, I know but it is confusing, you have to look on the code to see what this entity is. Would this redundancy be a problem?

Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py:16]: It was unstructured before so i suppose it's still better if it's redundant

**Review by Arcticae — APPROVED** [review]
I have no further requests, after addressing Maks's review can be merge from my perspective.

File: protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py
Comment (Arcticae) [reply protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py]: @karol-bisztyga please submit an issue to do that. it's a `good-first-issue` as well IMO :)

**Review by Arcticae — APPROVED** [review]
Mergable when Maks's comments are addressed.
Good job :)

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py:16]: Ok, so what's the final decision here? Can we go with this class? I don't mind that much though.

File: tests/integration/pure_cairo_vm/cheatcodes/expect_call/expect_call_with_proxy_test.cairo
Comment (karol-bisztyga) [reply tests/integration/pure_cairo_vm/cheatcodes/expect_call/expect_call_with_proxy_test.cairo:1]: Can we do this separately? I don't think it's so essential for now, honestly.

File: protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/callable_hint_locals/stop_expect_call_hint_local.py]: https://github.com/software-mansion/protostar/issues/1509

File: tests/integration/pure_cairo_vm/cheatcodes/expect_call/expect_call_with_proxy_test.cairo
Comment (karol-bisztyga) [reply tests/integration/pure_cairo_vm/cheatcodes/expect_call/expect_call_with_proxy_test.cairo:1]: https://github.com/software-mansion/protostar/issues/1510

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py]: [address fdbck](https://github.com/software-mansion/protostar/pull/1481/commits/6278d25d1bffdde6a7c5c54e89b6a5900221dba1)

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py]: [address fdbck](https://github.com/software-mansion/protostar/pull/1481/commits/6278d25d1bffdde6a7c5c54e89b6a5900221dba1)

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py:16]: ^this

File: tests/integration/pure_cairo_vm/cheatcodes/expect_call/expect_call_with_proxy_test.cairo
Comment (MaksymilianDemitraszek) [reply tests/integration/pure_cairo_vm/cheatcodes/expect_call/expect_call_with_proxy_test.cairo:1]: Ofc we can

Comment (MaksymilianDemitraszek) [reply tests/integration/pure_cairo_vm/cheatcodes/expect_call/expect_call_with_proxy_test.cairo:1]: But this is pretty important, we had a lot of problems with it historically

**Review by MaksymilianDemitraszek — APPROVED** [review]

File: protostar/cheatable_starknet/controllers/expect_call_controller.py
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/expect_call_controller.py]: [remove FunctionCall](https://github.com/software-mansion/protostar/pull/1481/commits/3aa4e7a7edcf559ea3e9593cac7d622633c23b8e)

**Pull request merged**
